### ISSUE TITLE
ISchemeHandler isn't multithreaded

### ISSUE THREAD

**Issue by davewthompson** [cefsharp/CefSharp#797]
I'm not sure if this is a bug or just the way that CefSharp is implemented, but it appears that when the browser invokes an implementation of *ISchemeHandler* to obtain a web resource this is done in a serial rather than a parallel fashion.

In my test ISchemeHandler implementation I'm immediately creating a Task and then calling the *requestCompletedCallback* on completion, but CefSharp never seems to call the *ProcessRequestAsync* method more than once before the first invocation returns.

Have I missed something? I'm using the standard versions. If this is by design then I apologise and please remove this bug.

Also, thanks for all your hard work!

**Comment by amaitland** [comment]
Can you provide a code sample? Preferably as a `GIST` or fork one of the `Example` projects.

I suspect there are some limitations imposed by `CEF`, though last time I looked at the `Scheme Handler` code it did look like it was a bit of a mess, so it's possible there's room for improvement.

I'm pretty sure the entry point is https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/SchemeHandlerWrapper.cpp#L33 should you wish to gain a better understanding of what's going on.

**Comment by rassilon** [comment]
I didn't see this happen when I made this ProcessRequestAsync method:

        public bool ProcessRequestAsync(IRequest request, ISchemeHandlerResponse response, OnRequestCompletedHandler requestCompletedCallback)
        {
            Uri uri = new Uri(request.Url);
            Debug.WriteLine("ProcessRequestAsync: " + request.Url);
            Task task = new Task(() =>
            {
                System.Threading.Thread.Sleep(200);
                response.ContentLength = 0;
                response.StatusCode = 200;
                response.CloseStream = true;
                response.ResponseStream = new MemoryStream();
                requestCompletedCallback.Invoke();
                Debug.WriteLine("ProcessRequestAsyncCompleted: " + uri.ToString());
            });
            task.Start();
            return true;
        }

Here's the debug output I observed:
> ProcessRequestAsync: URL1 
> ProcessRequestAsync: URL2 

~200ms gap

> ProcessRequestAsyncCompleted: URL1
> ProcessRequestAsyncCompleted: URL2

I caused this behavior by this HTML snippet (adjusted for the actual custom scheme name of course):
  
```
  <iframe id=id1></iframe>
  <iframe id=id2></iframe>
  <a href=javascript:testFunction()/>Blah Label</a>
  <script type="text/javascript">
      function testFunction()
      { id1.src = "customScheme://blah/blah"; id2.src = "customScheme://blah/blah2"; }
  </script>
```

and clicking on the `A` label.

So maybe this has more to do with how you're implementing the resource retrieval than the CefSharp side of things? A gist or fork of the Example projects is probably required for better understanding like @amaitland mentioned.

Fyi,
Bill

**Comment by rassilon** [comment]
@amaitland , upon further examination, `ClientAdapter::GetResourceHandler` is actually synchronous.  

This appears to be due to not using the CefSharp `SchemeHandlerWrapper` pattern of dealing with `CefResourceHandler`. This seems like an oversight.

Looking at CefSharp's `IResourceHandler`, this just looks like a naming mess.  

If I could break the world, I'd probably do something like this:
* Rename `IResourceHandler` to `IResourceHandlerManager` and `DefaultResourceHandler` to `DefaultResourceHandlerManager`.
* Consider moving `RegisterHandler/UnregisterHandler` to `ISimpleResourceHandlerManager` so they would be optional, but still implemented by `DefaultResourceHandlerManager`.
* Use ConcurrentDictionary for thread safety in the `Register/Unregister` methods.
* Rename `ISchemeHandler` to `IResourceHandler`.
* Rename `ProcessRequestAsync` to `ProcessResource`.
* Implement remaining members of `ISchemeResponse` on `ResourceHandler`.

This ought to allow async processing of `CefResourceHandler`s if desired.

I'm thinking that I won't be that lucky. :) So, I'll have to think some more about what a more backward compatible change might be. 

Any thoughts other folks might have would be appreciated.

Bill

**Comment by amaitland** [comment]
> This appears to be due to not using the CefSharp SchemeHandlerWrapper pattern of dealing with CefResourceHandler. This seems like an oversight.

@rassilon Can you provide a little more information? I don't have time to dig through code at the moment.

> This ought to allow async processing of CefResourceHandlers if desired.

Can you provide a little more information regarding the intent of your proposed changes? Most of them being renames, I'm unclear how this fixes the `async` problem?

> I'm thinking that I won't be that lucky. :)

In a major release like `39` I'm ok with breaking changes if they provide enough value to warrant them.

**Comment by rassilon** [comment]
Sure. Custom Schemes and GetResourceHandler both return CefResourceHandler subclasses to CEF.
Custom Schemes expose the callback parameter to allow CEF to be notified when a possible async operation is complete.

The current `IResourceHandler` isn't exposed this kind of detail, so you can only set the response stream and return. If you need to initiate a network connection it has to be on the CEF IO thread, blocking all other IO requests until it returns.

The intent is to be able to do the same things with IResourceHandler that you can do with ISchemeHandler.

That's the basic overview, the rest is the details of renaming the bits to more appropriate names given the underlying CefResourceHandler commonality.

Bill

**Comment by amaitland** [comment]
> The current IResourceHandler isn't exposed this kind of detail, so you can only set the response stream and return. If you need to initiate a network connection it has to be on the CEF IO thread, blocking all other IO requests until it returns.

I implemented `ResourceHandle` with the intention of loading mostly static content, intention wasn't to perform network requests. Though I do understand users will use the `API` however they wish.

If you keep the `CefSharp` public facing API changes to a minimal I don't see any problem with rewriting both. 

FYI you can register a `http` or `https` scheme handler and intercept any request you like and defer loading to `CEF` for the ones your not interested in handling.

**Comment by davewthompson** [comment]
Just a note, I use the custom handler to do network requests and I find it perfect for this - it works very well apart from the lack of asynchronous handling. I hijack HTTP requests to create a kind of in-process proxy.

Thank you for looking at this issue, it's much appreciated.

**Comment by rassilon** [comment]
@amaitland , How do you tell CEF to handle something a HTTP/HTTPS scheme handler doesn't want?

Just return false from your SchemeHandler if you don't want to handle it?

I thought that might do the trick, (I can try it again) but I switched to a HTTP/HTTPS scheme handler that only deals with an obviously fake DNS name: (native.appname kind of thing).

Before I used the fake DNS name so I could bind the handler to it, CEF didn't appear to handle the requests for me though.

Bill

**Comment by amaitland** [comment]
> Handlers can be used with both built-in schemes (HTTP, HTTPS, etc) and custom schemes

https://code.google.com/p/chromiumembedded/wiki/GeneralUsage#Scheme_Handler

It should just be a matter of `return false`, there maybe some other setting that's required. There are a number of settings in https://github.com/cefsharp/CefSharp/blob/master/CefSharp/CefCustomScheme.cs#L7

**Comment by amaitland** [comment]
> Before I used the fake DNS name so I could bind the handler to it, CEF didn't appear to handle the requests for me though.

If your using a fake dns name then you can set the `DomainName` on the scheme handler.

**Comment by rassilon** [comment]
Yeah, `DomainName` is what I'm currently using that works. I'll try returning `false` again tomorrow. It's probably not a big deal though.

I'm working on a CEF version of a native iOS app that currently embeds Safari on iOS.

The iOS app has this weird "don't trust the browser's caching" support that involves computing checksums of local .js/.css/.png files and asking the server to verify the checksums to see if new versions of the files need to be downloaded kind of a mechanic. If I have to use `IResourceHandler` for this kind of logic, it'd be nice for it to be async.

The iOS app/WinForms app both run in a kiosk like mode, so its not as if there's a reload button for the end user to try and fix things themselves.

Fyi,
Bill

**Comment by amaitland** [comment]
> I'll try returning false again tomorrow. It's probably not a big deal though.

Just read the API again `return false` will cancel the request. So I think it's just a combination of domain name and registering `http` or `https`. There was a `cefforum` post that I was reading the other day, can't seem to find it now though.

**Comment by amaitland** [comment]
Had a quick look at the code over lunch, I think it's possible to add `async` support to `ResourceHandler` without greatly affecting the `API`. Under the hood I'd be happy to see code cleanup, specially `SchemeHandlerWrapper`. 

- `IResourceHandler` should be renamed, it is confusing (my fault on that one). (I think it's useful for users to be directly able to register a handler without having to cast or implement their own version, that and it's used internally by `LoadHtml`, so I think the `register/unregister` methods can stay, they don't have to be implemented in a custom one).
- Add Thread Safety (I'm not super concerned about this one, as the user can implement it themselves, so I'd put this as low priority)
- Refactor `SchemeHandlerWrapper` to be generic enough for use in both scenarios, should be possible with delegates.
- Rename `SchemeHandlerWrapper` to `CefResourceHandlerWrapper` (seems like the most consistent name, as improvements are made, I'm trying to make things consistent overall.)
- Add `AsyncTaskResourceHandler`, then handle based on the type. Create a new one passing in a task using the constructor. Register callback using `ContinueWith`. (Rough idea, haven't looked into the implementation detail, may be just as easy to let the user specify their own `Task` or implement a default internal one, so there's only one code path).

Thoughts?

**Comment by amaitland** [comment]
Reading over this again and I remember having some problems with `SchemeHandler` and using a `Task` in the example app. From memory I was running the `WPF` example.

This is really just a reminder to myself to look into that further at some point.

https://github.com/cefsharp/CefSharp/commit/a0460e83d81bb7964d9027af2f69f277507443d1

**Comment by amaitland** [comment]
Linking this to #950

**Comment by amaitland** [comment]
Closed #950 and re-targeted to `2272` with #955
